### PR TITLE
GH-33985: [C++] Add substrait serialization/deserialization for expressions

### LINKED ISSUE #33985: [C++] Support specifying filters and projections with Substrait expressions
### Describe the enhancement requested

In addition to representing full plans, Substrait can also be used to represent expressions (see https://github.com/substrait-io/substrait/pull/405/). It would be nice if Dataset and Acero could consume Substrait expressions and use them to specify filters and projections. 

I would love to see us expose functions that:
- Receive a Boolean-valued Substrait scalar expression and use it to filter a Dataset
- Receive a list of Substrait scalar expressions and use them to Project a Dataset
- Receive a Boolean-valued Substrait scalar expression and use it to add a Filter node to the ExecPlan
- Receive a list of Substrait scalar expressions and use it to add a Project node to the ExecPlan

### Component(s)

C++

### PR DESCRIPTION
**Pull request by westonpace**
### Rationale for this change

Substrait provides a library-independent way to represent compute expressions.  By serializing and deserializing pyarrow compute expression to substrait we can allow interoperability with other libraries.

Originally it was thought this would not be needed because users would be sending entire query plans (which contain expressions) back and forth and so there was no need to work with expressions by themselves.

However, as more and more APIs and integration points emerge it turns out there are situations where serializing expressions by themselves is useful.  For example, the proposed datasets protocol, or for the Java JNI datasets implementation (which uses Arrow-C++'s datasets)

### What changes are included in this PR?

In Arrow-C++ we add two new methods to serialize and deserialize a collection of named, bound expressions to Substrait's ExtendedExpression message.

In pyarrow we expose these two methods and also add utility methods to pyarrow.compute.Expression to convert a single expression to/from substrait (these will be encoded as an ExtendedExpression message with one expression named "expression")

In addition, this PR exposed that we do not have very many bindings for arrow-functions to substrait-functions (previous work has mostly focused on the reverse direction).  This PR adds many (though not all) new bindings.

In addition, this PR adds ToProto for cast and both FromProto and ToProto support for the SingularOrList expression type (we convert is_in to SingularOrList and convert SingularOrList to an or list).

This should provide support for all the sargable operators except between (there is no Arrow-C++ between function) and like (we still don't have arrow->substrait bindings for the string functions) which should be a sufficient set of expressions for a first release.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

There are new features, as described above, but no backwards incompatible changes.

### Caveats

There are a fair number of minor inconsistencies or surprises, many of which can be smoothed over by follow-up work.

#### Bound Expressions

Arrow-C++ has long had a distinction between "unbound expressions" (e.g. `a + b`) and "bound expressions" (e.g. `a:i32 + b:i32`).  A bound expression is an expression that has been bound to a schema of some kind.  Field references are resolved and the output type is known for every node of the AST.

Pyarrow has hidden this complexity and most pyarrow compute expressions that the user encounters will be unbound expressions.  Substrait is only capable (currently) of representing bound expressions.  As a result, in order to serialize expressions, the user will need to provide an input schema.  This can be an inconvenience for some workflows.  To resolve this, I would like to eventually add support for unbound expressions to Substrait (https://github.com/substrait-io/substrait/issues/515)

Another minor annoyance of bound expressions is that an unbound pyarrow.compute.Expression object will not be equal to a bound pyarrow.compute.Expression object.  It would make testing easier if we had a `pyarrow.compute.Expression.equals` variant that did not examine bound fields.

#### Named field references

Pyarrow datasets users are used to working with named field references.  For example, one can set a filter `pc.equal(ds.field("x"), 7)`.  Substrait, since it requires everything to be bound, considers named references to be superfluous and does everything in terms of numeric indices into the base schema.  So the above expression, after round tripping, would become something like `pc.equal(ds.field(3), 7)` (assuming `"x"` is at index `3` in the schema used for serialization).  This is something that can be overcome in the future if Substrait adds support for unbound expressions.  Or, if that doesn't happen, it could still be implemented as a Substrait expression hint (this would allow named references to be used even if the user wants to work with bound expressions).

#### UDFs

UDFs ARE supported by this PR.  This covers both "builtin arrow functions that do not exist in substrait (e.g. shift_left)" and "custom UDFs added with `register_scalar_function`".  By default, UDFs will not be allowed when converting to Substrait because the resulting message would not be portable (e.g. you can't expect an external system to know about your custom UDFs).  However, you can set the `allow_udfs` flag to True and these will be allowed.  The Substrait representation will have the URI `urn:arrow:substrait_simple_extension_function`.

**Options**: Although UDFs are allowed we do not yet support UDFs that take function options.  These are trickier to convert to Substrait (though it should be possible in the future if someone is motivated enough).

#### Rough Edges

There are a few corner cases:

 * The function `is_in` converts to Substrait's `SingularOrList`.  On conversion back to Arrow this becomes an or list.  In other words, the function `is_in(5, [1, 2, 5])` converts to `5 == 1 || 5 == 2 || 5 == 5`.  This is because Substrait's or list is more expression and allows things like `5 == field_ref(0) || 5 == 7` which cannot be expressed as an `is_in` function.
 * Arrow functions can either be converted to Substrait or are considered UDFs.  However, there are a small number of functions which can "sometimes" be converted to Substrait depending on the function options.  At the moment I think this is only the `is_null` function.  The `is_null` function has an option `nan_is_null` which will allow you to consider `NaN` as a null value.  Substrait has no single function that evaluates both `NULL` and `NaN` as true.  In the meantime you can use `is_null || is_nan`.  In the future, should someone want to, they could add special logic to convert this case.
* Closes: #33985

### CODE REVIEW

**Comment by westonpace** [comment]
Leaving as draft as I need to add more test cases as well as python bindings.

**Comment by westonpace** [comment]
I've added python bindings.  Now all that is needed is documentation / examples

**Comment by westonpace** [comment]
I'm marking this ready for review.  I still want to add a few unit tests that verify we correctly raise errors when given Substrait expressions that Arrow cannot handle (e.g. MultiOrList) but these should be a minor addition.

**Comment by westonpace** [comment]
The appveyor failure seems valid (though utterly baffling): https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/47443264

File: python/pyarrow/tests/test_compute.py [hunk 1]
```diff
@@ -3277,19 +3281,106 @@ def test_expression_serialization():
...
+            serialized = expr.to_substrait(test_schema)
+            deserialized = pc.Expression.from_substrait(serialized)
+            serialized_again = deserialized.to_substrait(test_schema)
+            deserialized_again = pc.Expression.from_substrait(serialized_again)
+            assert deserialized.equals(deserialized_again)
+
+        # Special case, we lose the field names of struct literals
+        serialized = f.to_substrait(test_schema)
+        deserialized = pc.Expression.from_substrait(serialized)
+        assert deserialized.equals(pc.scalar({'': 1}))
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_compute.py:3371]: Should we use something else than an empty string? (eg "field0", "field1", ..) 
(although this is probably enough of a corner case to not put too much effort in it)

File: python/pyarrow/tests/test_compute.py [hunk 2]
```diff
@@ -3277,19 +3281,106 @@ def test_expression_serialization():
...
-                 pc.field('i64') == 7, pc.field('i64').is_null(),
-                 pc.field(('foo', 'bar')) == 'value',
-                 pc.field('foo', 'bar') == 'value']
+    all_exprs = literal_exprs + exprs_with_call + exprs_with_ref + special_cases
     for expr in all_exprs:
         assert isinstance(expr, pc.Expression)
         restored = pickle.loads(pickle.dumps(expr))
         assert expr.equals(restored)
 
+    if pas is not None:
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_compute.py:3319]: It _might_ be nicer to split out the substrait-based serialization to a separate test, but then we need to factor out the expression creation into a helper function? (it's certainly OK to just leave as is)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the nice and comprehensive description!

Did a review of the python bindings, which are looking good, just some minor comments.

**Comment by jorisvandenbossche** [comment]
> The function `is_in` converts to Substrait's `SingularOrList`.  On conversion back to Arrow this becomes an or list.  ....  This is because Substrait's or list is more expressive ..

And substrait doesn't have an "is_in" like function? (or are there plans for that?) 
(this conversion seems unfortunate, as "is_in" can be more efficient than the equivalent or-list)

File: python/pyarrow/_compute.pyx [hunk 3]
```diff
@@ -2350,6 +2367,58 @@ cdef class Expression(_Weakrefable):
...
+            If False then only functions that are part of the core Substrait function
+            definitions will be allowed.  Set this to True to allow pyarrow-specific functions
+            but the result may not be accepted by other compute libraries.
+
+        Returns
+        -------
+        Buffer
+            A buffer containing the serialized Protobuf plan.
+        """
+        return _pas().serialize_expressions([self], "expression", schema, allow_udfs=allow_udfs)
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_compute.pyx:2420]: Yeah, I was trying to hide the complexity for the to/from substrait methods since the substrait module could always be used for more advanced capabilities.

File: python/pyarrow/_substrait.pyx [hunk 4]
```diff
@@ -185,6 +187,143 @@ def _parse_json_plan(plan):
...
+    for i in range(len(exprs)):
+        if not isinstance(exprs[i], Expression):
+            raise TypeError(f"Expected Expression, got '{type(exprs[i])}'")
+        c_named_expr.expression = (<Expression> exprs[i]).unwrap()
+        if i < len(names):
+            if not isinstance(names[i], str):
+                raise TypeError(f"Expected str, got '{type(names[i])}'")
+            c_named_expr.name = tobytes(<str> names[i])
+        else:
+            c_named_expr.name = tobytes("autoname")
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_substrait.pyx:238]: Ok, I was on the fence here.  I don't have any real use case for it.  I agree it feels better to just give an error if they don't give the same number of names.  I'll fix this.

File: python/pyarrow/_substrait.pyx [hunk 5]
```diff
@@ -185,6 +187,143 @@ def _parse_json_plan(plan):
...
+
+    Parameters
+    ----------
+    exprs : list of Expression
+        The expressions to serialize
+    names : list of str
+        Names for the expressions
+    schema : Schema
+        The schema the expressions will be bound to
+    allow_udfs : bool, default False
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_substrait.pyx:212]: Ah, I forgot I needed to reconcile this :)  Both made sense to me.

Setting this to true enables both "arrow builtin functions that are not substrait functions" and "user registered udfs (or actual UDFs)" (from Substrait's perspective these are the same thing).  My thinking was that, as Substrait's function support expands, the first case wouldn't be encountered as much.  However, I think both names are still ok.  I'm happy to switch to "allow_arrow_extensions"

File: python/pyarrow/_substrait.pyx [hunk 6]
```diff
@@ -185,6 +187,143 @@ def _parse_json_plan(plan):
...
+        else:
+            c_named_expr.name = tobytes("autoname")
+        c_bound_exprs.named_expressions.push_back(c_named_expr)
+    c_bound_exprs.schema = (<Schema> schema).sp_schema
+
+    c_conversion_options.allow_arrow_extensions = allow_udfs
+
+    c_res_buffer = SerializeExpressions(c_bound_exprs, c_conversion_options)
+    with nogil:
+        c_buffer = GetResultValue(c_res_buffer)
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_substrait.pyx:246]: Probably just copy/paste :).  Yes, I will update this.

File: python/pyarrow/_substrait.pyx [hunk 7]
```diff
@@ -145,6 +146,7 @@ def run_query(plan, *, table_provider=None, use_threads=True):
         }
         c_conversion_options.named_table_provider = BindFunction[CNamedTableProvider](
             &_create_named_table_provider, named_table_args)
+        c_conversion_options.allow_arrow_extensions = False
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_substrait.pyx:149]: Yes, this is the default.  I can leave it off.

File: python/pyarrow/tests/test_compute.py [hunk 8]
```diff
@@ -3277,19 +3281,106 @@ def test_expression_serialization():
...
+            assert expr.equals(deserialized)
+
+        # Expressions are bound when they get serialized.  Since bound
+        # expressions are not equal to their unbound variants we cannot
+        # compare the round tripped with the original
+        for expr in exprs_with_call:
+            serialized = expr.to_substrait(test_schema)
+            deserialized = pc.Expression.from_substrait(serialized)
+            # We can't compare the expressions themselves because of the bound
+            # unbound difference. But we can compare the string representation
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/tests/test_compute.py:3338]: When we bind expressions, in addition to looking up field references, we look up matching kernels (and will actually apply implicit casts) and we store the matching kernel as part of the expression.  So "1 + 3" goes from:

```
function: add
args:
  1: i64
  3: i64
kernel: nullptr
```

...to...

```
function: add
args:
  1: i64
  3: i64
kernel: add<i64,i64>
```

And unfortunately, this means that Expression::Equals does not compare the two as equal.  I've added https://github.com/apache/arrow/issues/36427 to hopefully add that option at some point.

File: python/pyarrow/tests/test_compute.py (hunk 1 above)
Comment (westonpace) [reply python/pyarrow/tests/test_compute.py:3371]: Hmm, I think I'd rather leave it this way for now so it's more obvious that the fields were lost and not that they got mixed up with some other schema.  This is a bit of an obnoxious corner case.  We might want to fix this on the Substrait side.

**Comment by westonpace** [comment]
> And substrait doesn't have an "is_in" like function? (or are there plans for that?)
>  (this conversion seems unfortunate, as "is_in" can be more efficient than the equivalent or-list)

It's an interesting point.  We have things like this outside of expressions too.  For example, the "join" node doesn't distinguish between an equality join (which can be done efficiently with a hashmap) and a non-equality join (which cannot).  In that case we actually have both representations.  The one people typically use is the "JoinRel" which is a logical operator and thus allowed to be more generic without concern for efficiency and the other one is the "HashJoinRel" which is more specific / physical, but typically not created by producers (instead planners or optimizers convert from one to the other).

I think this is interesting because "is_in" vs. "singular-or-list" is basically a logical vs physical distinction for expressions which I don't think I've really considered before, but I agree with you its valid.

In any case, it will be easy enough in Acero's converter, to recognize the cases that can collapse to `is_in` and use it where appropriate.  I've created https://github.com/apache/arrow/issues/36535 to track this.

File: python/pyarrow/tests/test_compute.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply python/pyarrow/tests/test_compute.py:3338]: I see, thanks for the explanation! I hadn't considered that the function vs kernel distinction

**Comment by westonpace** [comment]
I also created https://github.com/substrait-io/substrait/issues/517 on the substrait side in case anyone wants to chime in.

**Comment by westonpace** [comment]
Hmmm...seems I've botched git.  Apologies.  Will fix.

File: python/pyarrow/_substrait.pyx (hunk 4 above)
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_substrait.pyx]: I've updated this to now raise an error if the length of names and exprs doesn't match (and I use zip now and added a test case).

File: python/pyarrow/_compute.pyx (hunk 3 above)
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_compute.pyx]: Oops.  This also gets caught now because exprs and names don't have the same length.  I've added a test for this as well.

File: python/pyarrow/_substrait.pyx [hunk 9]
```diff
@@ -185,6 +187,143 @@ def _parse_json_plan(plan):
     return pyarrow_wrap_buffer(c_buf_plan)
 
 
+def serialize_expressions(exprs, names, schema, *, allow_udfs=False):
+    """
+    Serialize a collection of expressions into Substrait
+
+    Substrait expressions must be bound to a schema.  For example,
+    the Substrait expression ``a_i32 + b_i32`` is different from the
+    Substrait expression ``a_i64 + b_i64``.  Pyarrow expressions are
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_substrait.pyx]: I've changed to this.

File: python/pyarrow/_substrait.pyx (hunk 5 above)
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_substrait.pyx]: I switched to `allow_arrow_extensions`

File: python/pyarrow/_substrait.pyx (hunk 6 above)
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_substrait.pyx]: Done.

File: python/pyarrow/_substrait.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_substrait.pyx]: I've removed this statement.

File: python/pyarrow/_substrait.pyx [hunk 10]
```diff
@@ -185,6 +187,143 @@ def _parse_json_plan(plan):
...
+        c_buffer = pyarrow_unwrap_buffer(py_buffer(buf))
+    elif isinstance(buf, Buffer):
+        c_buffer = pyarrow_unwrap_buffer(buf)
+    else:
+        raise TypeError(
+            f"Expected 'pyarrow.Buffer' or bytes, got '{type(buf)}'")
+
+    c_res_bound_exprs = DeserializeExpressions(deref(c_buffer))
+    with nogil:
+        c_bound_exprs = GetResultValue(c_res_bound_exprs)
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/_substrait.pyx]: I've moved this inside the nogil block.

File: python/pyarrow/tests/test_substrait.py [hunk 11]
```diff
@@ -923,3 +925,58 @@ def table_provider(names, _):
...
+    ])
+    exprs = [pc.equal(ds.field("x"), 7), pc.equal(ds.field("x"), ds.field("y"))]
+    buf = pa.substrait.serialize_expressions(exprs, ["first", "second"], schema)
+    returned = pa.substrait.deserialize_expressions(buf)
+    assert schema == returned.schema
+    assert len(returned.expressions) == 2
+
+    norm_exprs = [pc.equal(ds.field(0), 7), pc.equal(ds.field(0), ds.field(1))]
+    assert str(returned.expressions["first"]) == str(norm_exprs[0])
+    assert str(returned.expressions["second"]) == str(norm_exprs[1])
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/pyarrow/tests/test_substrait.py:981]: As mentioned below this doesn't work because of the kernels binding.

File: python/pyarrow/tests/test_compute.py (hunk 2 above)
Comment (westonpace) [reply python/pyarrow/tests/test_compute.py]: I went ahead and split them out.

**Comment by westonpace** [comment]
I've fixed the appveyor issue.  The remaining CI failures seem to be unrelated.  @jorisvandenbossche I don't know if you'll be able to give this another review?  I would like to get this into the release so people can start experimenting with it.

**Comment by westonpace** [comment]
I've rebased.  The one test failure is a GCS failure and unrelated.  Gentle ping for another review.

**Comment by westonpace** [comment]
I've rebased.  Friendly ping @jorisvandenbossche if you have time to review this.  If not, I plan on merging at the end of the week if things are still green.

**Comment by ianmcook** [comment]
@westonpace @jorisvandenbossche is this ready to merge?

**Comment by westonpace** [comment]
Yes.  Sorry, I have rebased one last time and will merge as soon as CI is green.

**Comment by westonpace** [comment]
Failures appear unrelated.

**Comment by kou** [comment]
@github-actions crossbow submit example-python-minimal-build-fedora-conda

File: python/pyarrow/tests/test_substrait.py [hunk 12]
```diff
@@ -21,8 +21,10 @@
 import pytest
 
 import pyarrow as pa
+import pyarrow.compute as pc
+import pyarrow.dataset as ds
```
Comment (kou) [inline python/pyarrow/tests/test_substrait.py:25]: @westonpace It seems that we can't import `pyarrow.dataset` unconditionally: https://github.com/ursacomputing/crossbow/actions/runs/5957902192/job/16161350103#step:3:8002

```text
==================================== ERRORS ====================================
_______________ ERROR collecting pyarrow/tests/test_substrait.py _______________
ImportError while importing test module '/arrow/python/pyarrow/tests/test_substrait.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
pyarrow/tests/test_substrait.py:25: in <module>
    import pyarrow.dataset as ds
pyarrow/dataset.py:58: in <module>
    raise ImportError(
E   ImportError: The pyarrow installation is not built with support for 'dataset' (No module named 'pyarrow._dataset')
```

Comment (kou) [reply python/pyarrow/tests/test_substrait.py:25]: Thanks!

**Comment by ianmcook** [comment]
In case this helps anyone, there's an example here showing how this can be used through Python: https://gist.github.com/ianmcook/f70fc185d29ae97bdf85ffe0378c68e0
